### PR TITLE
Add web.route-prefix flag

### LINKED ISSUE #73: -web.telemetry-path not honored for pushes
Hi,

I'm using pushgateway behind a proxy served by /push/*.

For this I run:

```
/srv/pushgateway -web.telemetry-path=/push/metrics
```

This works just fine:

```
$ wget -O - http://localhost:9091/metrics | wc -l
2016-01-26 22:07:15 ERROR 404: Not Found.
0
$ wget -O - http://localhost:9091/push/metrics | wc -l
2016-01-26 22:07:21 (1.20 GB/s) - written to stdout [6777/6777]
118
```

But for pushing this is not the same:

```
$ echo "some_metric 3.14" | curl --data-binary @-  http://localhost:9091/metrics/job/some_job
$
$ echo "some_metric 3.14" | curl --data-binary @-  http://localhost:9091/push/metrics/job/some_job
404 page not found
```

### PR DESCRIPTION
**Pull request by hairyhenderson**
Fixes #73 and supersedes #112 

I took a different approach - somewhat less elegant than originally proposed in #112. I'm open to suggestions 😉

I named the flag `-web.route-prefix` instead of `-web.external-url`, since it seemed the former is closer to what I think we're looking for in terms of how they're defined already in `prometheus`.

An aside - I had to dig into the git history to figure out how to do the `bindata.go` generation - seems the `bindata.go` Makefile target got removed in a previous commit, not sure if that was intentional. The command I used was:

```console
$ go-bindata -prefix=resources resources/...
```

Let me know how this looks - I'm happy to make any tweaks required!

### CODE REVIEW

**Comment by beorn7** [comment]
On my review queue.

**Comment by beorn7** [comment]
Apologies for the delay. I had a busy week. I'm on it now.

**Comment by beorn7** [comment]
BTW: The bindata Make target is now called `assets`. But I think you did the right thing. Essentially what `make assets` would have done anyway. (If you run it, you always get some changes because the file timestamps are baked into the bindata.)

File: main.go [hunk 1]
```diff
@@ -38,6 +38,7 @@ var (
 	showVersion         = flag.Bool("version", false, "Print version information.")
 	listenAddress       = flag.String("web.listen-address", ":9091", "Address to listen on for the web interface, API, and telemetry.")
 	metricsPath         = flag.String("web.telemetry-path", "/metrics", "Path under which to expose metrics.")
+	routePrefix         = flag.String("web.route-prefix", "", "Prefix for the internal routes of web endpoints.")
```
Comment (beorn7) [inline main.go]: Yes, definitely a good choice as it is in line with the current flag of the Prometheus server.

File: main.go [hunk 2]
```diff
@@ -49,13 +50,21 @@ func init() {
 func main() {
 	flag.Parse()
 
+	if *routePrefix == "/" {
+		*routePrefix = ""
+	}
+	if *routePrefix != "" && (*routePrefix)[0:1] != "/" {
+		log.Fatalf("-web.route-prefix value must begin with `/` (was `%s`)", *routePrefix)
+	}
```
Comment (beorn7) [inline main.go]: This requirement is inconsistent with the Prometheus server, which silently prepends the `/`. I think we can do the same here. Suggestion:

```Go
if *routePrefix != "" {
	*routePrefix = "/" + strings.Trim(*routePrefix, "/")
}
```

File: main.go [hunk 3]
```diff
@@ -73,30 +82,38 @@ func main() {
 	r.Handler("GET", *metricsPath, prometheus.Handler())
 
 	// Handlers for pushing and deleting metrics.
-	r.PUT("/metrics/job/:job/*labels", handler.Push(ms, true))
-	r.POST("/metrics/job/:job/*labels", handler.Push(ms, false))
-	r.DELETE("/metrics/job/:job/*labels", handler.Delete(ms))
-	r.PUT("/metrics/job/:job", handler.Push(ms, true))
-	r.POST("/metrics/job/:job", handler.Push(ms, false))
-	r.DELETE("/metrics/job/:job", handler.Delete(ms))
+	pushAPIPath := *routePrefix + "/metrics"
```
Comment (beorn7) [inline main.go]: At this point, we have already registered "/-/healthy", "/-/ready", and "/metrics". I think we want them prefixed as well. (At least that's what the Prometheus server is doing.) WDYT?

File: main.go [hunk 4]
```diff
@@ -73,30 +82,38 @@ func main() {
...
+	r.DELETE(pushAPIPath+"/jobs/:job", handler.LegacyDelete(ms))
+
+	r.Handler("GET", *routePrefix+"/static/*filepath", prometheus.InstrumentHandler(
 		"static",
 		http.FileServer(
-			&assetfs.AssetFS{Asset: Asset, AssetDir: AssetDir, AssetInfo: AssetInfo},
+			&assetfs.AssetFS{
+				Asset: func(name string) ([]byte, error) {
+					// Note: this will do bad things if routePrefix is multi-byte,
+					// but I don't think that's valid in a URL anyway.
```
Comment (beorn7) [inline main.go]: That's actually fine. `len` on a string acts on bytes, not runes, and the slice operations also act on the byte level.

File: main.go [hunk 5]
```diff
@@ -73,30 +82,38 @@ func main() {
...
 		),
 	))
 	statusHandler := prometheus.InstrumentHandlerFunc("status", handler.Status(ms, Asset, flags))
-	r.Handler("GET", "/status", statusHandler)
-	r.Handler("GET", "/", statusHandler)
+	r.Handler("GET", *routePrefix+"/status", statusHandler)
+	r.Handler("GET", *routePrefix+"/", statusHandler)
 
 	// Re-enable pprof.
 	r.GET("/debug/pprof/*pprof", handlePprof)
```
Comment (beorn7) [inline main.go]: We probably want the routing prefix here as well?

**Comment by beorn7** [comment]
About the `-web.external-url`: I don't think we need that in Pushgateway because we do not create any absolute URLs containing the host etc.

**Comment by hairyhenderson** [comment]
> However, the "github.com/prometheus/common/route" seems to have solved the problem of a prefixing router (based on "github.com/julienschmidt/httprouter") already. Perhaps you could use that one here?

I started out with that one, but I hit some roadblocks... I'll dig back into this and try to remember what those actually _were_ 😉

File: main.go (hunk 5 above)
Comment (hairyhenderson) [reply main.go]: My thought with `/debug` is that it's not necessarily something you want to expose to the same audience as the other routes. I suppose this is something that can be blocked by reverse proxies, though.

I'll prefix this for now.

File: main.go (hunk 2 above)
Comment (hairyhenderson) [reply main.go]: Ok, seems reasonable. I'll try that out.

**Comment by brian-brazil** [comment]
> About the -web.external-url: I don't think we need that in Pushgateway because we do not create any absolute URLs containing the host etc.

If we don't need that, do we need web.route-prefix? The prefix was added as an extra configuration option for when you were using an external URL.

**Comment by beorn7** [comment]
@brian-brazil See https://github.com/prometheus/pushgateway/pull/146 for the use case. As the PGW never needs to know about its own hostname in URLs, just the route prefix should do the trick.

File: main.go (hunk 3 above)
Comment (hairyhenderson) [reply main.go]: > At this point, we have already registered "/-/healthy", "/-/ready", and "/metrics". I think we want them prefixed as well. (At least that's what the Prometheus server is doing.) WDYT?

I've prefixed `/-/*` with `*routePrefix`, but it seems like `*metricsPath` should be unaffected, since it's configured separately.

File: main.go (hunk 4 above)
Comment (hairyhenderson) [reply main.go]: 🤔 I thought I understood this... I remember going back and forth on this very late at night and I can't remember now why I thought it could cause a problem 🤷‍♂️

**Comment by hairyhenderson** [comment]
@beorn7 I've updated with most of the comments addressed - I'm still looking into `github.com/prometheus/common/route`, and may update this later on if it makes sense.

**Comment by hairyhenderson** [comment]
> I started out with [github.com/prometheus/common/route], but I hit some roadblocks...

Ok, figured out what the roadblocks were. It's mostly around the fact that I'd need to do a bunch of type-casting and wrapping, because `handler.Push()` returns a `httprouter.Handle`, and not a `http.HandlerFunc`. If https://github.com/prometheus/common/blob/master/route/route.go#L42 were exported, it'd be much easier 😉

I'm going to leave this as-is for now, and maybe some day it can be refactored to use `github.com/prometheus/common/route`.

**Comment by hairyhenderson** [comment]
> If https://github.com/prometheus/common/blob/master/route/route.go#L42 were exported, it'd be much easier

I take it back - it's the other way around... not as easy to adapt that way.

**Comment by hairyhenderson** [comment]
All that to say, @beorn7 this is ready to re-review 😉

**Comment by beorn7** [comment]
Thanks. On my review queue (again).

**Comment by beorn7** [comment]
Still on my queue. Sorry for the delay.

**Comment by hairyhenderson** [comment]
@beorn7 no worries - thanks for the update 🙂

File: main.go (hunk 3 above)
Comment (beorn7) [reply main.go]: Or we could prefix the configured metricsPath, or we can prefix it only if metricsPath is not configured explicitly. But in practice, we should just do what the Prometheus server is doing. Interestingly, at some point it was decided that we don't even need a `--web.telemetry-path` flag anymore for the Prometheus server. But before that flag was removed, the `--web.route-prefix` was used to prefix also the `--web.telemetry-path`. That's one point for doing it in the PGW, too. The other point is that we most likely will eventually remove `--web.telemetry-path` from the PGW, too, In that case, we would definitely need to prefix the (hardcoded) `/metrics` path. Which means we should do it now, too.

Long story short: Please prefix the metrics exposition path, too. Configured or not.

**Comment by hairyhenderson** [comment]
@beorn7 PTAL again - I've addressed the last remaining comment

File: main.go [hunk 6]
```diff
@@ -67,39 +77,45 @@ func main() {
 	// prometheus.EnableCollectChecks(true)
 
 	r := httprouter.New()
-	r.Handler("GET", "/-/healthy", prometheus.InstrumentHandlerFunc("healthy", handler.Healthy(ms)))
-	r.Handler("GET", "/-/ready", prometheus.InstrumentHandlerFunc("ready", handler.Ready(ms)))
+	r.Handler("GET", *routePrefix+"/-/healthy", prometheus.InstrumentHandlerFunc("healthy", handler.Healthy(ms)))
+	r.Handler("GET", *routePrefix+"/-/ready", prometheus.InstrumentHandlerFunc("ready", handler.Ready(ms)))
 
-	r.Handler("GET", *metricsPath, prometheus.Handler())
+	r.Handler("GET", *routePrefix+*metricsPath, prometheus.Handler())
```
Comment (beorn7) [inline main.go]: If `metricsPath` doesn't have a leading `/`, and a `routePrefix` is set, this will result in something weird. Easy way out: Use https://golang.org/pkg/path/#Join

**Comment by beorn7** [comment]
Which has triggered one last follow-up. See above.

File: main.go (hunk 6 above)
Comment (hairyhenderson) [reply main.go]: 🤦‍♂️ - I _had_ used `path.Join`, but I guess I didn't commit that 😉

**Comment by hairyhenderson** [comment]
Ok, should be good to go now.

**Comment by beorn7** [comment]
Thanks again. Merging now…

**Comment by hairyhenderson** [comment]
Thanks @beorn7 🙂

Is there a 0.4.1 or 0.5.0 in the cards any time soon? I see #143 though it seems there's a dependency on #90 being fixed first...

**Comment by beorn7** [comment]
Yeah, that's a pretty nasty bug I would not like to see in a release again. It's quite high on my priority list, but sadly, there are many items still higher.
